csi/controller.go

```diff
@@ -371,6 +392,63 @@ func (s *OsdCsiServer) DeleteVolume(
 	return &csi.DeleteVolumeResponse{}, nil
 }
 
+// ControllerExpandVolume is a CSI API which resizes a volume
+func (s *OsdCsiServer) ControllerExpandVolume(
+	ctx context.Context,
+	req *csi.ControllerExpandVolumeRequest,
+) (*csi.ControllerExpandVolumeResponse, error) {
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	} else if req.GetCapacityRange() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Capacity range must be provided")
+	} else if req.GetCapacityRange().GetRequiredBytes() < 0 || req.GetCapacityRange().GetLimitBytes() < 0 {
+		return nil, status.Error(codes.InvalidArgument, "Capacity ranges values cannot be negative")
+	}
+
+	// Get parameters
+	spec := &api.VolumeSpecUpdate{}
+
+	// Get Size
+	newSize := uint64(req.GetCapacityRange().GetRequiredBytes())
+	if newSize != 0 {
+		spec.SizeOpt = &api.VolumeSpecUpdate_Size{
+			Size: newSize,
+		}
+	}
+
+	// Get grpc connection
+	conn, err := s.getConn()
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Internal,
+			"Unable to connect to SDK server: %v", err)
+	}
+
+	// Get secret if any was passed
+	ctx = s.setupContextWithToken(ctx, req.GetSecrets())
+
+	// If the new size is greater than the current size, a volume update
+	// should be issued. Otherwise, no operation should occur.
+	volumes := api.NewOpenStorageVolumeClient(conn)
+
+	// Update volume with new size
+	_, err = volumes.Update(ctx, &api.SdkVolumeUpdateRequest{
```

Comment:
Originally, I was checking if the current volume size is less than the new size (in order to verify an expand was needed), but `volumes.Update` checks for this and provides a good error message:

```
 pxctl v update 1050284155922952603 -s 100
Update Volume: Failed to update volume: Error updating size on remote node: Shrinking of volumes is not supported. Current size for volume 6df869038d7f39d899f998f6abb0a553b634018d58e7bec1996046d14284c32c is 101 GiB
```